Ackquire AIK during attestation

arturkow2000 opened the pull request.

arturkow2000 committed:
Prepare for attestation

Refactor code, move common parts into shared modules.

arturkow2000 committed:
Move more code into common parts

arturkow2000 committed:
Fix warnings in client/provisioning, add attestation client

Add dummy attestation client. Implement logic to switch between
provisioning and attestation. When starting Fobnail checks whether it
has any platform provisioned and if that's the case enters attestation
mode, otherwise enters provisioning mode.

arturkow2000 committed:
Fix clippy warnings.

arturkow2000 committed:
Remove manual release of trussed borrow

There is no need to release trussed manually. Currently we don't use
Trussed neither during provisioning or attestation due to lack of RSA
support.

arturkow2000 committed:
Small fixes

Fix a few typos, remove do_verify_metadata_signature method and call
decode_signed_object directly.

arturkow2000 committed:
Drop Ed25519 support

We don't use Ed25519 anymore as most TPMs don't implement it. Ed25519
code is dead for a long time already and there is no need to still keep
it.

arturkow2000 committed:
Save AIK to persistent storage

Save AIK to persistent to load it from storage during attestation
instead of fetching it again each time.

arturkow2000 committed:
Load AIK from internal storage during attestation.

Previously we were receiving AIK during attestation and doing complete
EK recv, verification, credential activation procedure, however this
approach had security issues and wasn't conformant with RATS.

arturkow2000 committed:
Remove unused method from certmgr

The only thing len() is doing, is generating unnecessary warnings.

arturkow2000 committed:
Nonce support

To prevent replay attacks during attestation Fobnail generates random 32
byte nonce, which is sent as part of request. Attester appends this
nonce before computing reponse signature.

arturkow2000 committed:
Verify metadata signature after loading AIK

krystian-hebel commented on src/client/crypto.rs:
```diff
@@ -81,9 +21,24 @@ impl RsaKey<'_> {
         }
     }
+
+    pub fn load2(n: &[u8], e: &[u8]) -> Result<Self, ()> {
```
Is it possible to store `e` as a slice so `load` can be used to load key from file, too?

arturkow2000 replied:
@krystian-hebel You mean to store as u32? Yes, it is possible.

arturkow2000 replied:
Fixed

krystian-hebel commented on src/client/signing.rs:
```diff
@@ -105,3 +57,23 @@ where
+        error!("Failed to deserialize signed object (outer): {}", e);
+    })?;
+
+    let sha = trussed::try_syscall!(trussed.hash(
```
Is this the only place where `MAX_MESSAGE_LENGTH` limits us?

It is sad that Trussed still uses `update` under the hood (https://github.com/trussed-dev/trussed/blob/main/src/mechanisms/sha256.rs#L43) but doesn't expose it. It should be possible to implement in a stateless way, by passing internal state (same size as hash, plus maybe additional integer for keeping size info) for each function...

arturkow2000 replied:
@krystian-hebel It limits file size too

arturkow2000 replied:
Increased limit to 8192 and tested whether still works on nRF.

arturkow2000 committed:
Correct an error message

We check signature of metadata (response), not request.

arturkow2000 committed:
Change format for AIK saved in persistent storage

Use the same encoding as used for transmitting AIK between attester and
Fobnail (exponent encoded as u32 instead of byte array).

arturkow2000 committed:
Increase Trussed MAX_MESSAGE_LENGTH, don't panic when message is too big.
